Libraries/LibCore/Promise.h

```diff
@@ -24,6 +24,42 @@ class Promise : public EventReceiver {
     Function<ErrorOr<void>(Result&)> on_resolution;
     Function<void(ErrorType&)> on_rejection;
 
+    static NonnullRefPtr<Promise> after(Vector<NonnullRefPtr<Promise>>&& promises)
+    {
+        auto promise = Promise::construct();
+        struct Resolved : RefCounted<Resolved> {
+            explicit Resolved(size_t n)
+                : needed(n)
+            {
+            }
+
+            size_t count { 0 };
+            size_t needed { 0 };
+        };
+
+        auto resolved = make_ref_counted<Resolved>(promises.size());
+        auto weak_promise = promise->template make_weak_ptr<Promise>();
+        for (auto p : promises) {
+            p->when_resolved([weak_promise, resolved](Result&) -> ErrorOr<void> {
```

Comment:
this use of weak promises without checking them makes me nervous, but I do understand that the strong ref edge is from the 'returned' promise to its children via add_child.

Does calling `WeakPtr<T>::operator->` create a strong ref each time it's called? or does it short-circuit and assert that it's not dead before going in.

Reply:
Neither actually, it just returns the raw ptr, but it's guaranteed to be alive if the current promise is alive.